bigframes/core/rewrite.py

```diff
@@ -311,3 +326,25 @@ def get_node_column_ids(node: nodes.BigFrameNode) -> Tuple[str, ...]:
     import bigframes.core
 
     return tuple(bigframes.core.ArrayValue(node).column_ids)
+
+
+def common_subtree(
+    l_tree: nodes.BigFrameNode, r_tree: nodes.BigFrameNode
+) -> Optional[nodes.BigFrameNode]:
+    """Find common subtree between join subtrees"""
```

Comment:
To find common substree between two trees, the substree should have same node type and as well all children nodes should be same as well. IIUC, the below code can return a substree when its root node are same while its child nodes can be different. Another question here, is the `REWRITABLE_NODE_TYPES` unary nodes only? Can be binary node also?

Reply:
equality check is deep by default with dataclasses
try this for example. this returns `True`
```import dataclasses

@dataclasses.dataclass(frozen=True)
class Bar:
  a: int
  b: int

@dataclasses.dataclass(frozen=True)
class Foo:
  a: Bar


Foo(Bar(1,2)) == Foo(Bar(1,2))
```

Also, `REWRITABLE_NODE_TYPES` are the set of nodes that can be rewritten from a join into a projection.